# Домашнее задание 3: Обучение LLaMA

**Мягкий дедлайн:** неделя с момента получения (5 декабря 2024)

---

## Задание 1: LLM для генерации текста

### **Цель:**  
Ознакомиться с использованием языковых моделей для генерации текста.

### **Описание задачи:**  
Напишите код для генерации текстов с использованием предобученной языковой модели Hugging Face, например, GPT-2 или другой подходящей модели. Генерируйте текст по заданной теме, например:  
- "Прогноз погоды"
- "Советы по фитнесу"
- "История из будущего"

### **Ключевые шаги:**  

1. Установите и настройте библиотеку Hugging Face Transformers.
2. Загрузите предобученную модель (например, GPT-2).
3. Напишите функцию для генерации текста на основе текстового префикса.
4. Проверьте работу модели с разными параметрами (например, длина текста, температура, топ-k).
5. Сравните результаты для разных начальных префиксов.

### **Рекомендации:**  

- **Базовый вариант:** Использовать GPT-2.
- **Лучший подход:** Изучите модели LLaMA или Mistral, которые можно запустить в Google Colab (скорее всего, не выше 8B параметров).

---

## Задание 2: Классификация текста с BERT

### **Цель:**  
Понять, как использовать предобученные модели типа BERT для задач классификации текста.

### **Описание задачи:**  
Реализуйте модель классификации текстов на основе предобученной модели BERT. Используйте датасет [IMDb Reviews](https://www.kaggle.com/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews) для классификации отзывов на положительные и отрицательные, оцените модель и сделайте выводы.  

- **Подходы:**  
  - Можно дообучить модель на этом датасете.
  - Можно использовать zero-shot классификацию.  

Главное — добиться нужного качества классификации.

---

### **Дополнительные баллы (альтернатива заданию 2):**

- Выберите модель на русском языке.
- Проведите классификацию запросов, где задача определить, нужно ли отправить фото или нет.  

Пример:  
- "Скинь фото" → 1  
- "Как дела?" → 0  

- **Задача:** Использовать zero-shot классификацию.  
- **Задачи с русским языком:** Найдите подходящую модель, добавьте в блокнот примеры вызова.

---


### Задача 1

Параметры: 
temperature=0.5

max_length=50

top_k=50

top_p=0.95

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# модель GPT-2
model_name = "gpt2" 
device = "cuda" if torch.cuda.is_available() else "cpu"

#загрузка модели и токенизатора
print("Загружаем модель...")
tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(model_name).to(device)
print("Модель загружена.")

#генерация текста
def generate_text(prefix, max_new_tokens=50, temperature=0.7, top_k=50, top_p=0.9):
    input_ids = tokenizer.encode(prefix, return_tensors="pt").to(device)
    output = model.generate(
        input_ids,
        max_new_tokens=max_new_tokens,  
        temperature=temperature,
        top_k=top_k,
        top_p=top_p,
        do_sample=True,
        pad_token_id=tokenizer.eos_token_id
    )
    return tokenizer.decode(output[0], skip_special_tokens=True)

#префиксы
topics = [
    "Weather forecast: Today it is expected to be",
    "Fitness tips: To achieve the best results",
    "Story from the future: In the year 2142, humanity finally"
]

print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=100, temperature=1.0, top_k=50, top_p=0.9))
    print("-" * 50)


Загружаем модель...
Модель загружена.

Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy and snow-covered. The main river that connects the river to the town is expected to be water. In the morning and evening the river will be very cold but should cool by noon.

There is a large amount of rainfall. The rainfall will be low but the temperature will still be above freezing and a good view is available.

Snow

The rainy days usually last over a year. But there are a few major storms that hit the country on a daily basis. The
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results, follow your diet and exercise routine. Follow up with your doctor to help you find the foods you need to lose weight.

Read or Share this story: http://usat.ly/29jH7Pt
--------------------------------------------------
Тема: Story from the future: In 

Параметры: 
temperature=0.7

max_length=100

top_k=30

top_p=0.95

In [42]:
print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=100, temperature=0.7, top_k=30, top_p=0.9))
    print("-" * 50)



Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy with a slight breeze. The clouds will be very thick and will reach 100°F.

Tonight's forecast is expected to be cloudy with a slight breeze. The clouds will be very thick and will reach 100°F.

The weather forecast: The weather forecast for the day of the 11th is below. It will be cloudy with a slight breeze. The clouds will be very thick and will reach 100°F.

The weather forecast: The weather forecast for the day
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results, be sure to keep your body warm, hydrated and hydrated. Don't worry if you lose weight. Keep your weight in the gym.

You can also use the following diet tips:

Eat less than 30% of your calories from fat and carbs

Eat less than 15% of your calories from protein and fat

Eat less than 30% of your calories fro

Параметры: 
temperature=0.4

max_length=30

top_k=10

top_p=0.95

In [ ]:

print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=30, temperature=0.4, top_k=10, top_p=0.95))
    print("-" * 50)



Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy.

The weather forecast for the day is:

The weather forecast for the day is:

The weather forecast for the day
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results, you need to be able to perform at least one of the following:

1. Perform a full-body workout

2. Perform
--------------------------------------------------
Тема: Story from the future: In the year 2142, humanity finally
Story from the future: In the year 2142, humanity finally reaches the end of its existence.

The story follows the story of the first human to reach the end of its existence, and the story of
--------------------------------------------------


Лучше всего посмотреть как влияют гиперпараметры на одинаковой длине текста. Пусть будет max_length=30

Параметры: 
temperature=0.1

top_p=0.95

top_k=10

In [46]:

print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=30, temperature=0.1, top_k=10, top_p=0.95))
    print("-" * 50)


Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy and cloudy with a low of -5C.

The weather service said the wind was blowing at a high rate of speed and the wind
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results, you need to be able to perform at least one of the following:

1. Perform a full-body workout

2. Perform
--------------------------------------------------
Тема: Story from the future: In the year 2142, humanity finally
Story from the future: In the year 2142, humanity finally reached the point where it could no longer be controlled by the forces of nature. The only way to stop this was to destroy the planet.


--------------------------------------------------


Параметры: 
temperature=0.1

top_p=0.95

top_k=1

In [48]:
print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=30, temperature=0.1, top_k=1, top_p=0.95))
    print("-" * 50)


Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy.

The weather forecast for the day is:

The forecast for the day is:

The forecast for the day is:
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results, you need to be able to perform at least one of the following:

1. Perform a full-body workout

2. Perform
--------------------------------------------------
Тема: Story from the future: In the year 2142, humanity finally
Story from the future: In the year 2142, humanity finally has a chance to stop the plague.

The story of the year 2142 is a story about a group of people who have been living in
--------------------------------------------------


Видно, что смена top_k= с 10 на 1 сделала более строгий отбор топовых токенов, что снизило креативность выхода модели

Параметры: 
temperature=1.5

top_p=0.95

top_k=1

In [49]:
print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=30, temperature=1.5, top_k=10, top_p=0.95))
    print("-" * 50)


Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cold in all major regions. The coldest part of the day is around 10°C with a low humidity at 2°F and high humidity at
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results from training, make sure that each exercise is performed at a consistent pace, using your bodyweight.

If you're trying to avoid cardio,
--------------------------------------------------
Тема: Story from the future: In the year 2142, humanity finally
Story from the future: In the year 2142, humanity finally conquered all life.

The world is still not quite the same

The human race is still living on a planet without any human life and
--------------------------------------------------


Видно, что смена temperature= с 0.1 на 1.5 сделала более креативный вывод модели.

Параметры: 
temperature=1.5

top_p=0.4

top_k=1

In [50]:
print("\nСгенерированные тексты:\n")
for topic in topics:
    print(f"Тема: {topic}")
    print(generate_text(prefix=topic, max_new_tokens=30, temperature=1.5, top_k=10, top_p=0.4))
    print("-" * 50)


Сгенерированные тексты:

Тема: Weather forecast: Today it is expected to be
Weather forecast: Today it is expected to be cloudy and cloudy with a few showers and a few showers in the afternoon.

Tonight's forecast is cloudy and cloudy with a few showers and a
--------------------------------------------------
Тема: Fitness tips: To achieve the best results
Fitness tips: To achieve the best results in the gym, you need to get more out of your workouts. If you're looking for a way to increase your fitness, you need to get
--------------------------------------------------
Тема: Story from the future: In the year 2142, humanity finally
Story from the future: In the year 2142, humanity finally reached a point where it could not be stopped, and the planet was in a state of war.

The story of the story is one of
--------------------------------------------------


Как мы видим, уменьшение top_p с 0.95 на 0.4 ухудшило качество классификации, например, по топику погоды видно, что слова повторяются

Как вывод, с помощью temperature и top_k мы можем влиять на креативность модели, но нужно понимать, что чем сильнее креативность генерации, тем выше вероятность получить какой-то странный бесмысленный текст. Длиной текста мы можем влиять на , как нистранно, длину генерации. Однако, тут нужно понимать, что чем больше будет длина, тем выше переятность получить бессмысленный текст. А вероятность выбора токена top_p контролирует выбор следующего токена на основе совокупной вероятности и низкое значение может ухудшить результат.

### Задача 2

In [ ]:
import os
import pandas as pd
import torch
from torch.utils.data import Dataset, DataLoader
from transformers import BertTokenizer, BertForSequenceClassification, AdamW, get_linear_schedule_with_warmup
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from tqdm.auto import tqdm

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Используется устройство:", device)



#датасет
data_path = r"E:\HSE\ФКН\NLP\HW3\IMDB Dataset.csv"
data = pd.read_csv(data_path)

#меняем'positive'/'negative' вна 1 и 0 
data['label'] = data['sentiment'].apply(lambda x: 1 if x == 'positive' else 0)

reviews = data['review'].tolist()
labels = data['label'].tolist()


train_texts, test_texts, train_labels, test_labels = train_test_split(
    reviews, labels, test_size=0.2, random_state=42, stratify=labels
)



#токенизация

model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name, do_lower_case=True)
max_length = 256

class IMDBDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_length=256):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        text = self.texts[idx]
        label = self.labels[idx]

        encoding = self.tokenizer(
            text,
            return_tensors='pt',
            padding='max_length',
            truncation=True,
            max_length=self.max_length
        )
        return {
            'input_ids': encoding['input_ids'].squeeze(0),
            'attention_mask': encoding['attention_mask'].squeeze(0),
            'labels': torch.tensor(label, dtype=torch.long)
        }

train_dataset = IMDBDataset(train_texts, train_labels, tokenizer, max_length=max_length)
test_dataset = IMDBDataset(test_texts, test_labels, tokenizer, max_length=max_length)

batch_size = 16
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)


# инициируем BERT 

model = BertForSequenceClassification.from_pretrained(model_name, num_labels=2)
model.to(device)


# гиперпараметры

epochs = 2
optimizer = AdamW(model.parameters(), lr=2e-5)
total_steps = len(train_loader) * epochs
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=total_steps)


#обучаем модель


def train_epoch(model, dataloader):
    model.train()
    total_loss = 0
    for batch in tqdm(dataloader, desc="Обучение"):
        optimizer.zero_grad()

        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)

        outputs = model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=labels
        )
        loss = outputs.loss
        total_loss += loss.item()

        loss.backward()
        optimizer.step()
        scheduler.step()

    avg_loss = total_loss / len(dataloader)
    return avg_loss

def evaluate(model, dataloader):
    model.eval()
    preds, true_labels = [], []
    with torch.no_grad():
        for batch in tqdm(dataloader, desc="Оценка"):
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)

            outputs = model(input_ids=input_ids, attention_mask=attention_mask)
            logits = outputs.logits

            predictions = torch.argmax(logits, dim=-1)
            preds.extend(predictions.cpu().numpy())
            true_labels.extend(labels.cpu().numpy())

    acc = accuracy_score(true_labels, preds)
    return acc


#обучение и оценка

for epoch in range(epochs):
    print(f"Эпоха {epoch+1}/{epochs}")
    train_loss = train_epoch(model, train_loader)
    print(f"Средняя ошибка за эпоху: {train_loss:.4f}")

    test_acc = evaluate(model, test_loader)
    print(f"Точность на тестовой выборке: {test_acc:.4f}")


# accuracy_score

test_acc = evaluate(model, test_loader)
print(f"\nИтоговая точность на тестовом наборе: {test_acc:.4f}")


Используется устройство: cpu


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

c:\Anaconda\Lib\site-packages\huggingface_hub\file_download.py:139: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Никита\.cache\huggingface\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Эпоха 1/2


c:\Anaconda\Lib\site-packages\transformers\optimization.py:591: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Обучение:   0%|          | 0/2500 [00:00<?, ?it/s]

Средняя ошибка за эпоху: 0.2426


Оценка:   0%|          | 0/625 [00:00<?, ?it/s]

Точность на тестовой выборке: 0.9240
Эпоха 2/2


Обучение:   0%|          | 0/2500 [00:00<?, ?it/s]

Средняя ошибка за эпоху: 0.1178


Оценка:   0%|          | 0/625 [00:00<?, ?it/s]

Точность на тестовой выборке: 0.9279


Оценка:   0%|          | 0/625 [00:00<?, ?it/s]


Итоговая точность на тестовом наборе: 0.9279


Мы дообучили модель BERT на нашем датасете, Accuracy оказался 92.79%, что говорит о том, что модель хорошо справляется с задачей классификации отзывов